src/interactive-window/interactiveWindow.ts

```diff
@@ -110,33 +107,85 @@ export class InteractiveWindow implements IInteractiveWindowLoadable {
     } = {};
     private pendingNotebookScrolls: NotebookRange[] = [];
 
+    private _notebookEditor!: NotebookEditor;
+    public get notebookEditor(): NotebookEditor {
+        return this._notebookEditor;
+    }
+
+    private _notebookUri: Uri;
+    public get notebookUri(): Uri {
+        return this._notebookUri;
+    }
+
+    private readonly documentManager: IDocumentManager;
+    private readonly fs: IFileSystem;
+    private readonly configuration: IConfigurationService;
+    private readonly jupyterExporter: INotebookExporter;
+    private readonly workspaceService: IWorkspaceService;
+    private readonly exportDialog: IExportDialog;
+    private readonly notebookControllerSelection: IControllerSelection;
+    private readonly interactiveWindowDebugger: IInteractiveWindowDebugger | undefined;
+    private readonly errorHandler: IDataScienceErrorHandler;
+    private readonly codeGeneratorFactory: ICodeGeneratorFactory;
+    private readonly storageFactory: IGeneratedCodeStorageFactory;
+    private readonly debuggingManager: IInteractiveWindowDebuggingManager;
+    private readonly isWebExtension: boolean;
+    private readonly commandManager: ICommandManager;
     constructor(
-        private readonly documentManager: IDocumentManager,
-        private readonly fs: IFileSystem,
-        private readonly configuration: IConfigurationService,
-        private readonly commandManager: ICommandManager,
-        private readonly jupyterExporter: INotebookExporter,
-        private readonly workspaceService: IWorkspaceService,
-        private _owner: Resource,
-        private mode: InteractiveWindowMode,
-        private readonly exportDialog: IExportDialog,
-        private readonly notebookControllerSelection: IControllerSelection,
         private readonly serviceContainer: IServiceContainer,
-        private readonly interactiveWindowDebugger: IInteractiveWindowDebugger | undefined,
-        private readonly errorHandler: IDataScienceErrorHandler,
-        preferredController: IVSCodeNotebookController | undefined,
-        public readonly notebookEditor: NotebookEditor,
-        public readonly inputUri: Uri,
-        public readonly appShell: IApplicationShell,
-        private readonly codeGeneratorFactory: ICodeGeneratorFactory,
-        private readonly storageFactory: IGeneratedCodeStorageFactory,
-        private readonly debuggingManager: IInteractiveWindowDebuggingManager,
-        private readonly isWebExtension: boolean
+        private _owner: Resource,
+        public mode: InteractiveWindowMode,
+        private preferredController: IVSCodeNotebookController | undefined,
+        private readonly notebookEditorOrTab: NotebookEditor | InteractiveTab,
+        public readonly inputUri: Uri
     ) {
+        this.documentManager = this.serviceContainer.get<IDocumentManager>(IDocumentManager);
```

Comment:
Why did you change all of these to come from the service container? 

This has caused problems in the past when we ship because sometimes the items aren't available anymore in the service container (somebody refactors something somewhere else). With an explicit dependency this is impossible.

Reply:
Sorry I take that back. This was already doing that. You basically moved this from the provider to here.